File: 프로그래머스/lv3/118668. 코딩 테스트 공부/코딩 테스트 공부.py
import sys
sys.setrecursionlimit(10**8)
input = sys.stdin.readline

INF = 987654321
dp = [[-1] * 200 for _ in range(200)]
problems = []
max_alp = -1
max_cop = -1

def solve(alp, cop):
    if alp >= max_alp and cop >= max_cop:
        return 0

    if dp[alp][cop] != -1:
        return dp[alp][cop]

    dp[alp][cop] = INF

    dp[alp][cop] = min(solve(min(max_alp, alp + 1), cop) + 1, solve(alp, min(max_cop, cop + 1)) + 1)
    for i in range(len(problems)):
        if alp >= problems[i][0] and cop >= problems[i][1]:
            dp[alp][cop] = min(dp[alp][cop], solve(min(max_alp, alp + problems[i][2]), min(max_cop, cop + problems[i][3])) + problems[i][4])
    return dp[alp][cop]

def solution(alp, cop, problem):
    global max_alp, max_cop
    global problems
    problems = problem
    global  dp

    for i in range(len(problems)):
        max_alp = max(max_alp, problems[i][0])
        max_cop = max(max_cop, problems[i][1])

    return solve(alp, cop)

File: 프로그래머스/3/118668. 코딩 테스트 공부/코딩 테스트 공부.java
import java.util.*;

class Solution {
    public static int maxAlp = -1;
    public static int maxCop = -1;
    public static int[][] dp = new int[1001][1001];
    public static int INF = 987654321;
    public int solution(int alp, int cop, int[][] problems) {
      for (int[] problem : problems) {
            maxAlp = Math.max(maxAlp, problem[0]);
            maxCop = Math.max(maxCop, problem[1]);
        }

        dp = new int[Math.max(alp, maxAlp) + 1][Math.max(cop, maxCop) + 1];

        for (int i = 0; i <= Math.max(alp, maxAlp); i++) {
            Arrays.fill(dp[i], INF);
        }

        return solve(alp, cop, problems);
    }
    
    static int solve(int alp, int cop, int[][] problems){
        if(alp >= maxAlp && cop >= maxCop){
            return 0;
        }
        if(dp[alp][cop] != INF){
            return dp[alp][cop];
        }
        dp[alp][cop] = INF + 1;
        for(int i = 0; i < problems.length; ++i){
            if(alp >= problems[i][0] && cop >= problems[i][1]){
                dp[alp][cop] = Math.min(dp[alp][cop], problems[i][4] + solve(Math.min(alp + problems[i][2], maxAlp), Math.min(cop + problems[i][3], maxCop), problems));
            }
        }
        
        dp[alp][cop] = Math.min(dp[alp][cop], 1 + solve(Math.min(alp + 1, maxAlp), cop, problems));
        dp[alp][cop] = Math.min(dp[alp][cop], 1 + solve(alp, Math.min(cop + 1, maxCop), problems));

        return dp[alp][cop];
    }
}